src/doc/en/installation/conda.rst

```diff
@@ -126,22 +129,22 @@ Here we assume that you are using a git checkout.
   - Create and activate a new conda environment with the dependencies of Sage
     and a few additional developer tools::
 
-      $ mamba env create --file src/environment-dev.yml --name sage-dev
+      $ mamba env create --file src/environment-dev-3.11.yml --name sage-dev
       $ conda activate sage-dev
 
-    Alternatively, you can use ``src/environment.yml`` or
-    ``src/environment-optional.yml``, which will only install standard
+    Alternatively, you can use ``src/environment-3.11.yml`` or
+    ``src/environment-optional-3.11.yml``, which will only install standard
     (and optional) packages without any additional developer tools.
 
-    By default, the most recent version of Python supported by Sage is
-    installed. You can use the additional option ``python=3.9`` in the above
-    ``env create`` command to select another Python version (here 3.9). 
+    A different Python version can be selected by replacing ``3.11`` by ``3.9``
+    or ``3.10`` in these commands.
 
   - Run the ``configure`` script::
 
       $ ./bootstrap
       $ ./configure --with-python=$CONDA_PREFIX/bin/python             \
                     --prefix=$CONDA_PREFIX                             \
+                    --enable-system-site-packages                      \
```

Comment:
Why is this needed / what is it doing in this context? Also since the word 'system' package might be misleading for conda, and the list of additional commands is getting rather big, I propose you introduce a new toggle "with-conda=$CONDA_PREFIX" to configure that takes care of all this config.

Reply:
> what is it doing in this context?

It's unfortunately not documented yet (@mjo, am I right about this?), and is as experimental as this conda installation method.

Here I am using it so that the version constraints of those Python packages with spkg-configure are checked. Extra checks = good.

How long this command is does not really matter -- it is to be copy-pasted anyway. In any case, inventing new switches for conda is beyond the scope of this PR.

Reply:
Is it also changing how the venv is setup/used?

> Here I am using it so that the version constraints of those Python packages with spkg-configure are checked. Extra checks = good.

I don't see the point in that. We have ci checks already in place and the lock files I proposed would make it even more unnecessary.

Reply:
> Is it also changing how the venv is setup/used?

Note that in these instructions, no venv is created.

Reply:
An important point of running the `configure` script with the "force" options is to check that the installed versions are suitable for Sage. It is of course just a partial check because it is only done for those package for which we have the spkg-configure.m4. Now thanks to @orlitzky's work, we can check more, so here I'm adding it. It's a natural improvement.

Reply:
I know, but my understanding of this toggle is that it tries to break out of the venv and uses system installed packages. So does this mean it only uses packages that are installed in the conda environment, or also ones that are installed by a system python?

Reply:
I'll also note that in updating the branch, synchronizing the version constraints on the conda side with what is recorded in install-requires.txt (and was updated in package updates) was a major element. This extra check would have been valuable for validating these versions.

Reply:
The conda-lock business from the stale PR #35986 is irrelevant for this discussion. Someone has to create the lock files and make sure that they are OK. So even if all users just use the lockfiles, the maintainer of the lockfiles will benefit from the extra checks.

Reply:
> my understanding of this toggle is that it tries to break out of the venv and uses system installed packages. So does this mean it only uses packages that are installed in the conda environment, or also ones that are installed by a system python?

Here, the "system" is the conda environment. This is consistent with how we talk about "system packages" elsewhere.

Reply:
> > what is it doing in this context?
> 
> It's unfortunately not documented yet (@mjo, am I right about this?), and is as experimental as this conda installation method.

It's not documented beyond the `./configure --help` description.

Basically, it's the first step towards treating python packages like any other. By passing `--enable-system-site-packages`, 

 1. You allow python to see your system site packages, i.e. `pyvenv.cfg` gets `include-system-site-packages = true`.
 2. You allow `./configure` to detect any python packages that it can see and that meet the version constraints declared in `install-requires.txt`.

It does not detect "user" site packages yet, and is disabled by default. It's experimental because it turns on a huge number of system packages and the only way to be sure that our version constraints are accurate is to have people actually try them. Particularly, distribution (and conda, nix, etc.) developers should give it a try because it serves as an early warning system for version incompatibilities.

Reply:
> > my understanding of this toggle is that it tries to break out of the venv and uses system installed packages. So does this mean it only uses packages that are installed in the conda environment, or also ones that are installed by a system python?
> 
> Here, the "system" is the conda environment. This is consistent with how we talk about "system packages" elsewhere.

Also, I stole the name from `python -m venv`.

Reply:
Thanks for the clarification @orlitzky.
Concerning 1), I'm right in the assumption that this is not relevant here as we don't create a venv when used with conda.
Concerning 2), what happens if the python package doesn't meet the version constraints?

I'm totally fine with using this new feature. I'm just trying to understand what additional value we get beyond the already existing version checks in https://github.com/sagemath/sage/blob/1cf0c13e5271b13f0db4364a39d8bd80e13f66cc/.github/workflows/ci-conda.yml#L117-L120?

Reply:
Thanks @orlitzky; I've taken the liberty to put an edited version of this in https://github.com/sagemath/sage/wiki/Sage-10.2-Release-Tour#new-developer-tools-modularization-deprecations

Reply:
> Concerning 1), I'm right in the assumption that this is not relevant here as we don't create a venv when used with conda.

I think the `--enable-system-site-packages` flag was only added to the **Using conda to provide all dependencies for the Sage library (experimental)** scenario, where sage is built "normally" and conda is used only for the dependencies. In that case, I think you still get a venv.

> Concerning 2), what happens if the python package doesn't meet the version constraints?

`./configure` rejects the system package and you install the SPKG instead.

Reply:
> Concerning 1), I'm right in the assumption that this is not relevant here as we don't create a venv when used with conda. 

Yes, that's right.

> Concerning 2), what happens if the python package doesn't meet the version constraints?

If `--with-system-SPKG=force` has been used for a package, it is an error. Otherwise, it is not an error.

Reply:
> I'm just trying to understand what additional value we get beyond the already existing version checks in

`pip check` only sees the versions and version constraints recorded by the installed distribution packages. Only some of the information in our `build/pkgs/*/install-requires.txt` files ends up in there -- namely the information from those packages that are listed in `src/setup.cfg` (generated from `src/setup.cfg.m4`).

Reply:
> and then during `pip install src` pip would install a updated version of that package.

also in this step, the version ranges only applies to those packages that are listed in src/setup.cfg (generated from src/setup.cfg.m4).

Reply:
> The new behavior is that `configure` fails now. Why is this advantageous?

This flags the inconsistency to developers, so it can be fixed.

Reply:
Then it should be in the conda test, but not as a recommended way in the docs. (maybe as a "optionally, you can add the flag ..." sentence afterwards if you think it provides enough value).

Reply:
Likely you've forgotten that even when configure gives an error from the package tests, it still writes out the config files; so users can continue with the next step of the installation regardless.
`--enable-system-site-packages` really is the correct switch to use for this mode of installation -- the Python packages provided in the conda environment are to be inspected by `configure`. It improves the output of `configure`, as it brings the summary lines ("system package will be used" / "will be installed") closer to the truth.

Reply:
The point of conda is to move as much of responsibility to conda, and use as little as possible from sage's own build system. This change here is counterproductive towards this goal.

If there are issues with consistency for the moment, then sure, please add this as tests to the conda workflow. But we also don't recommend that users check with "pip check" afterwards, that everything went okay.

Reply:
The configuration step won't go away. This has nothing to do with using or not using the Sage build system. (We aren't using it in this mode, as you know.)

Reply:
> Concerning 1), I'm right in the assumption that this is not relevant here as we don't create a venv when used with conda.

I can provide a better answer to this now.

Right: it's not directly relevant to the conda installation. But it _does_ provide a sanity check on the compatibility of the versions used by conda and sage. In the long run, that will help conda.

We are currently adding `spkg-configure.m4` files for all of sage's python packages, testing the version bounds in `install-requires.txt`, adding compatibility patches, etc. When all that is done, and when enough people have tested it, my hope is that we can turn `--enable-system-site-packages` on by default. If that happens, using conda to install dependencies for sage becomes much simpler because you don't have to circumvent `make` and the venv.

For that to work smoothly, the conda versions all have to be acceptable to `./configure`; otherwise, `./configure` will either crash or want to install an SPKG. Adding `--enable-system-site-packages` now, while not having any direct effect, does flag those incompatibilities to us.

Reply:
> using conda to install dependencies for sage becomes much simpler because you don't have to circumvent make

It's actually a feature and not a bug that the conda install doesn't use the make interface. One of the longterm goals is to replace sages package management and build infrastructure with conda. Right now this is already working very well for python packages. You could remove every python package (but keep the auto generated conda env and setup.cfg files) and everything is still working. That's why I think this flag here is a step backwards (for the conda install, I'm sure it has other good usages).

I would propose we remove this flag from the docs for now, but keep it in the ci workflow. Then you do get the consistency checks, just that not everyone has to run them.

Reply:
I don't think anyone is proposing to add the use of `make` to the all-conda installation method.

Reply:
That was also not my point...